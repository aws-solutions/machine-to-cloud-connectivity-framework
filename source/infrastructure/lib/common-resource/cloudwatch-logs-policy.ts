// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { ArnFormat, Stack } from 'aws-cdk-lib';
import { Effect, PolicyDocument, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';

/**
 * Creates a common CloudWatch Logs policy for Lambda functions.
 */
export class CloudwatchLogsPolicyConstruct extends Construct {
  public policy: PolicyDocument;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    this.policy = new PolicyDocument({
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents'],
          resources: [
            Stack.of(this).formatArn({
              service: 'logs',
              resource: 'log-group',
              resourceName: '/aws/lambda/*',
              arnFormat: ArnFormat.COLON_RESOURCE_NAME
            })
          ]
        })
      ]
    });
  }
}
