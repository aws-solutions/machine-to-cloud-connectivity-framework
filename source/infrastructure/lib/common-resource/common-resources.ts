// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { ArnFormat, RemovalPolicy, Stack } from 'aws-cdk-lib';
import { AnyPrincipal, Effect, PolicyDocument, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { BlockPublicAccess, Bucket, BucketAccessControl, BucketEncryption, IBucket } from 'aws-cdk-lib/aws-s3';
import { NagSuppressions } from 'cdk-nag';
import { Construct } from 'constructs';
import { addCfnSuppressRules } from '../../utils/utils';

export interface CommonResourcesConstructProps {
  sourceCodeBucket: string;
}

/**
 * Creates a common CloudWatch Logs policy for Lambda functions and a logging S3 bucket.
 */
export class CommonResourcesConstruct extends Construct {
  public cloudWatchLogsPolicy: PolicyDocument;
  public s3LoggingBucket: Bucket;
  public sourceCodeBucket: IBucket;

  constructor(scope: Construct, id: string, props: CommonResourcesConstructProps) {
    super(scope, id);

    this.cloudWatchLogsPolicy = new PolicyDocument({
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents'],
          resources: [
            Stack.of(this).formatArn({
              service: 'logs',
              resource: 'log-group',
              resourceName: '/aws/lambda/*',
              arnFormat: ArnFormat.COLON_RESOURCE_NAME
            })
          ]
        })
      ]
    });

    this.s3LoggingBucket = new Bucket(this, 'LogBucket', {
      accessControl: BucketAccessControl.LOG_DELIVERY_WRITE,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      encryption: BucketEncryption.S3_MANAGED,
      removalPolicy: RemovalPolicy.RETAIN
    });
    this.s3LoggingBucket.addToResourcePolicy(
      new PolicyStatement({
        actions: ['s3:*'],
        conditions: {
          Bool: { 'aws:SecureTransport': 'false' }
        },
        effect: Effect.DENY,
        principals: [new AnyPrincipal()],
        resources: [this.s3LoggingBucket.bucketArn, this.s3LoggingBucket.arnForObjects('*')]
      })
    );
    addCfnSuppressRules(this.s3LoggingBucket, [
      { id: 'W35', reason: 'This bucket is to store S3 logs, so it does not require access logs.' }
    ]);

    this.sourceCodeBucket = Bucket.fromBucketName(this, 'SourceCodeBucket', props.sourceCodeBucket);

    // cdk-nag suppressions
    NagSuppressions.addResourceSuppressions(this.s3LoggingBucket, [
      { id: 'AwsSolutions-S1', reason: 'This bucket is to store S3 logs, so it does not require access logs.' }
    ]);
  }
}
