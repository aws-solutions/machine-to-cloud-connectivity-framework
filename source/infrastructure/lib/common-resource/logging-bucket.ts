// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { RemovalPolicy, Aws } from 'aws-cdk-lib';
import { AnyPrincipal, Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Bucket, BucketEncryption, ObjectOwnership } from 'aws-cdk-lib/aws-s3';
import { NagSuppressions } from 'cdk-nag';
import { Construct } from 'constructs';
import { addCfnSuppressRules } from '../../utils/utils';

/**
 * Creates a common CloudWatch Logs policy for Lambda functions.
 */
export class LoggingBucketConstruct extends Construct {
  public s3LoggingBucket: Bucket;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    this.s3LoggingBucket = new Bucket(this, 'LogBucket', {
      objectOwnership: ObjectOwnership.OBJECT_WRITER,
      encryption: BucketEncryption.S3_MANAGED,
      removalPolicy: RemovalPolicy.RETAIN,
      bucketName: `${Aws.STACK_NAME}-${Aws.ACCOUNT_ID}-log`
    });
    this.s3LoggingBucket.addToResourcePolicy(
      new PolicyStatement({
        actions: ['s3:*'],
        conditions: {
          Bool: { 'aws:SecureTransport': 'false' }
        },
        effect: Effect.DENY,
        principals: [new AnyPrincipal()],
        resources: [this.s3LoggingBucket.bucketArn, this.s3LoggingBucket.arnForObjects('*')]
      })
    );
    addCfnSuppressRules(this.s3LoggingBucket, [
      { id: 'W35', reason: 'This bucket is to store S3 logs, so it does not require access logs.' }
    ]);

    // cdk-nag suppressions
    NagSuppressions.addResourceSuppressions(this.s3LoggingBucket, [
      { id: 'AwsSolutions-S1', reason: 'This bucket is to store S3 logs, so it does not require access logs.' },
      { id: 'AwsSolutions-S2', reason: 'Public Access Blocking is handled by objectOwnership' }
    ]);
  }
}
