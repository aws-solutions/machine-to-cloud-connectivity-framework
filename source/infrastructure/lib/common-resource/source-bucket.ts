// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { IBucket, Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

export interface SourceBucketConstructProps {
  sourceCodeBucketName: string;
}

/**
 * Imports existing bucket containing source code
 */
export class SourceBucketConstruct extends Construct {
  public sourceCodeBucket: IBucket;

  constructor(scope: Construct, id: string, props: SourceBucketConstructProps) {
    super(scope, id);

    this.sourceCodeBucket = Bucket.fromBucketName(this, 'SourceCodeBucket', props.sourceCodeBucketName);
  }
}
