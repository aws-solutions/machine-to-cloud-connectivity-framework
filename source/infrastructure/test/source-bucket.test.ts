// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from 'aws-cdk-lib';
import { SourceBucketConstruct } from '../lib/common-resource/source-bucket';

test('M2C2 source bucket test', () => {
  const stack = new Stack();
  const sourceBucket = new SourceBucketConstruct(stack, 'LoggingBucket', {
    sourceCodeBucketName: 'test-name'
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  expect(sourceBucket.sourceCodeBucket).toBeDefined();
});
