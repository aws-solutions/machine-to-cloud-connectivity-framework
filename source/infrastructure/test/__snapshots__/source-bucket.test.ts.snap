// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`M2C2 source bucket test 1`] = `Object {}`;
